lambdaclass/curse_of_mirra
# GoogleSignIn - Iteration 1 - SignIn/SingOut/SignInWithCached - Android and iOS

#961 

## Motivation

Implement GoogleSignIn so players can link their google acoounts to the game.

## Summary of changes

- Added GoogleSignIn plugin
- Created `GoogleSignInController` script to controll the SingIn/SingOut/SingInWithCache
- Added funcionality to the Signin button in TittleScreen
- For now we are storing the google information in the PlayersPrefs. Further Iteration it will be stored in our DB.
- Implemented cancelation method when the request takes too long.
- Make it work in Android and iOS
- Created Google OAuth console credentials and app.
- Added https://github.com/googlesamples/unity-jar-resolver. To maange Android and Ios dependencies

Scripts inside GoogleSignInPlugin folder are part of the GoogleSignIn plugin I did not create them.

More information can be found in the following docs I created:

- [Google SignIn options and SingUp/SignIn sequence diagrams](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/BJvny1jO6)
- [Google SignIn OAauth console and unity config](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/HyRViKaOp)
- [GoogleSignIn iOS configuration](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/ByjGNWeq6)
- [Google OAuth set to production](https://hackmd.io/@2D_U0_CeRO-2SX-Q68fDnA/BJEV4iTtp)

## How has this been tested?

This can only be tested in mobile, whether it's a simulator (not unity simulator) or a real phone.

Setup:
1. Either ask a @BertovDev or @Nico-Sanchez to add you in their credentials or create your own.
2. Install JDK [here](https://download.oracle.com/java/22/latest/jdk-22_macos-aarch64_bin.dmg)
3. Replace the google ClientID in client/Assets/Scripts/GoogleSignIn/GoogleSignInController.cs:24
4. export the `GOOGLE_CLIENT_ID` variable with the google clientID (2) before running the backend.
5. Replace the url in client/Assets/Scripts/Utils/ServerUtils.cs:226 with ngrok's url pointing to `localhost:4001` while running backend app.

ℹ️ You can create your own testing credentials following this guide -> https://www.notion.so/lambdaclass/Google-Credentials-172a36447ed447dc9d26c202a0b0bf09

1. Press SignIn, a google popup should appear. Choose your account
6. Press play  you should see your google name displayed in the top left corner in Mainscreen
7. Close the App, re-open it. You should be logged with your last account entered.
8. SignOut. Now if you close and re-open no account should be appear as logged
9. SignIn again with any account you want.

Next steps:

- Add .env for credentials
- Handle Errors
- Relationate Google credentials to user in the DB
- How to re-authenticate the user data in the backend
- Change Domains

## Test additions / changes

List tests added/updated.

## Checklist
